Srpg 56 item base

jhwoo944 opened the pull request:
## 개요
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
아이템 (=인벤토리에 등록되는 객체)
초안 스크립트 작성 및 구조 설계
- SO를 상속하는 클래스 ItemData 생성
- 아이템 타입별로 SO스크립트를 작성
- 각각의 타입별 ItemSO 클래스는 모두 ItemData를 부모 클래스로 가짐.

현재 등록된 아이템 타입
- 티켓 (가챠 티켓)
- 조각 (캐릭터 조각)
- 토큰 (골드와 별개로 상점에서 교환 등에 쓰이는 토큰)
- 소비품 (AP충전제, 골드상자 등 Open & Reward가 있는 아이템)
- 캐릭터 돌파 소재 (기획에는 없지만 일단 구현해둠)
- 캐릭터 경험치 책
- 스킬 레벨업 소재
- 장비 돌파 소재 (기획에는 없지만 일단 구현해둠)
- 장비 경험치 결정
- 선물 (호감도)
- 기념품 (기념품)
//
- 아이템 / 아이템 SO 관련 폴더 생성 / 샘플 SO 추가 및 디렉터리 정리
- Constants에 enum: ItemType, ItemRank, Rewards 추가
//
AccounData에 다음 메서드를 추가 (인벤토리 최소 기능)
 - 아이템 추가 메서드
 - 아이템 제거 메서드
 - 인벤토리 출력 메서드 (Debug.Log)
 - 아이템 갯수 체크 / 차감 메서드
 - 아이템 확인 메서드
<!---- Resolves: #(Isuue Number) -->

## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [x] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다. (ex. feature/SRPG-11-UIBase)
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

jhwoo944 committed:
SRPG-56 ItemBase

 - ItemData 폴더 및 cs 파일 생성

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-56-ItemBase

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-56-ItemBase

jhwoo944 committed:
SRPG-56 Feat: ItemData 및 ItemSO 작성

 - 모든 아이템 타입의 부모가 되는 ItemData를 작성
 - 게임에 존재할 수 있는 아이템 분류군을 enum으로 작성
 - 각 아이템 분류군별로 SO 스크립트를 생성
 - SO스크립트별 내용 작성은 아직 미완성

jhwoo944 committed:
SRPG-56 SO 디렉터리 변경

 - 장비 / 아이템의 생성 메뉴의 세부 디렉터리 추가

jhwoo944 committed:
SRPG-56 Feat: 아이템SO 샘플 추가

아이템 (=인벤토리에서 등록되는 객체)
초안 스크립트 작성 및 구조 구축

 - SO를 상속하는 클래스 ItemData 생성
 - 아이템 타입별로 SO스크립트를 작성하고 샘플 SO 생성
 - 각각의 ItemSO 클래스는 모두 ItemData를 부모 클래스로 가짐.

현재 등록된 아이템 타입
 - 티켓 (가챠 티켓)
 - 조각 (캐릭터 조각)
 - 토큰 (골드와 별개로 상점에서 교환 등에 쓰이는 토큰)
 - 소비품 (AP충전제, 골드상자 등 Open & Reward가 있는 아이템)
 - 캐릭터 돌파 소재 (기획에는 없지만 일단 구현해둠)
 - 캐릭터 경험치 책
 - 스킬 레벨업 소재
 - 장비 돌파 소재 (기획에는 없지만 일단 구현해둠)
 - 장비 경험치 결정
 - 선물 (호감도)
 - 기념품 (기념품)
//
 - 아이템 / 아이템 SO 관련 폴더 생성 및 디렉터리 정리
 - Constants에 enum: ItemType, ItemRank, Rewards 추가

jhwoo944 committed:
SRPG-56 ItemBase: AccountData에 인벤토리 메서드추가

AccounData에 다음 메서드를 추가 (인벤토리 최소 기능)
 - 아이템 추가 메서드
 - 아이템 제거 메서드
 - 인벤토리 출력 메서드 (Debug.Log)
 - 아이템 갯수 체크 / 차감 메서드
 - 아이템 확인 메서드

jhwoo944 committed:
SRPG-56 Fix: 주석 수정

 - 주석 수정

jhwoo944 committed:
SRPG-29 Feat: CharacterGrowth 분리 작업

CharacterGrowth에서 CharacterGrowthManager를 분리
 - CharacterGrowth : 캐릭터의 개인 고유 성장 데이터를 저장함
 - CharacterGrowthManager : 캐릭터의 각종 스탯 및 필드를 계산 및 처리하는 메서드를 저장함
CharacterGrowthManager는 Managers에 등록되어 있으며, 모든 캐릭터가 공용으로 사용하고 Init으로 처리할 character를 받아옵니다.

 - 모든 SO(캐릭터/스킬/특성/클래스/아이템)의 ID 값 자료형을 int로 수정했습니다.
 - PlayerData의 UID값은 string인데 int로 변경할지 논의 필요.
 - 그 외 피드백 받은 부분 수정

SeRDic0705 commented on Assets/Scripts/Utility/Constants.cs:
```diff
@@ -141,4 +141,38 @@ public enum ArmorType //(방어구 종류는 임의로 설정함)
 
+    //아이템 분류.
+    //아이템 분류 선별 기준 = '인벤토리에 표시되고 정보를 볼 수 있어야 하는지? 여부가 성립하면 아이템'
+    public enum ItemType 
```
더미데이터는 일단 주석처리 해놓고 실제 사용할 때 해제하는 것이 좋을 것 같습니다

SeRDic0705 commented on Assets/Scripts/Objects/Accounts/AccountData.cs:
```diff
@@ -8,26 +9,117 @@ public class AccountData
     public PlayerData playerData { get; set; }
     public Dictionary<int, bool> missionClearData { get; set; }
-    //public Dictionary<ItemData, int> inventory { get; set; }  Todo: 아이템 데이터 추가 시 활성화 필요
+    public Dictionary<ItemData, int> inventory { get; set; }  //인벤토리 = 아이템 데이터, 보유 갯수
```
서버와의 입출력까지도 고려하면 다른 데이터값들과 마찬가지로 아이템별로 고유 id를 만들고 key값을 int로 사용하는 것이 좋습니다

jhwoo944 replied:
현재는 ItemData DB가 없어서 key를 id로 처리하면 Inventory 관련 메서드에서 ItemData 필드에 접근할 방법이 없어서 테스트에 지장이 있을 것 같아서 일단은 ItemData인 채로 두었습니다.

SeRDic0705 replied:
그렇가면 임시로 두고 추후에 변경하면 되겠네요
확인했습니다:)

SeRDic0705 commented on Assets/Scripts/Objects/Items/ItemData.cs:
```diff
@@ -0,0 +1,23 @@
+    public string itemDescription;//������ ����
+
+    [Header("Image")]
+    public Sprite itemSprite;//������ ��������Ʈ(=�����or������)
```
어제 회의 때 특성 스프라이트를 직접 가지고 있지 말고 Load해서 사용하자고 의견이 나왔던거 같은데

아이템이나 다른 데이터들도 마찬가지로 Load해서 사용하는 방식으로 통일하는게 좋을 것 같아요

jhwoo944 replied:
Load식으로 하자고 결정된 건 알지만 구체적으로 어떻게 Load할 지는 정해지지 않아서 일단 이대로 남겨뒀습니다.
경로값을 string으로 저장하도록 하면 될까요?

SeRDic0705 replied:
아이콘이미지를 클래스가 들고있는 상태로 호출하는 것이 아니라

아이콘을 표시해야할 때, 아이템의 id값을 통해 접근하는 것이기 때문에

id만 있으면 지워버려도 무방합니다

SeRDic0705 commented on Assets/Scripts/Objects/Items/Item_ConsumableData/ConsumableSO.cs:
```diff
@@ -0,0 +1,21 @@
+using static Constants;
+
+[CreateAssetMenu(menuName = "ItemData/ConsumableData", fileName = "Consumable_")]
+public class ConsumableSO : ItemData
```
Consumable아이템이라 체력회복포션 같은 걸 생각했는데, Ap나 골드교환권 같은 개념인가요?

보통 소비 아이템을 따로 만들어서 지급하기보다는 즉시 지급하는 경우가 많은데, 특별히 생각하신 이유가 있을까요?

jhwoo944 replied:
제가 한 게임들은 일단 AP 회복 아이템 / 골드 교환권은 거의 필수로 있고 그 외에도 이것저것 교환권 느낌의 아이템이 있는 경우가 많아서 이렇게 설정했습니다. 확장성을 생각하면 교환권 타입이 있는 게 좋을 것 같아요.

jhwoo944 replied:
추가로 인게임 배틀에서 아이템을 사용한다는 언급은 없어서 그 타입의 아이템은 고려를 안했습니다.
만약 인게임 배틀에서 아이템을 쓸 수 있게 된다면 밸런스 책정도 힘들어지고 게임 자체로도 재미를 해칠 것 같기도 하고요.

SeRDic0705 replied:
네 현우님의 생각이 맞습니다. 배틀아이템 사용으로 인해 밸런스 조정이 어려워질 수 있기에 기획단계에서 다음과 같은 사항을 고려했습니다.


- pvp 콘텐츠에서는 배틀아이템 사용이 불가
- pve 콘텐츠에서는? 확장성을 고려했을 때, 추후 보스레이드나 챌린지맵 등이 등장할 수 있어 배틀아이템의 필요성 자체는 인지
- 밸런스에 대해서는 한번의 스테이지에서 사용할 수 있는 수량 종류의 제한
- 겸사겸사 포션아이템 공방을 추가하여 장비아이템을 강화하는 재료의 소모처 추가

그러나 우리는 라이브서비스가 목표가 아니라 "프로젝트 개발 경험"이 목표이기 때문에, 개발 직관성에 초점을 맞춰 반드시 필요한 파트 외에는 가지치기를 했습니다.

전투시스템은 필수지만 배틀아이템은 있으나 없으나 게임에는 큰 상관이 없죠.
 비슷한 궤에서 캐릭터호감도 역시 생각만해두고 구현하지 않을 계획입니다.

물론 우리가 실력이 부족해서 구현 못하는게 아니죠. 하지만 주어진 일정이 부족하기 때문에 필수요소들을 완벽하게 진행하고 그 후에 확장을 고려해야 할 것 같아요.

그런 의미에서 일단 즉시 획득하게 개발하고, 추후에 수정기간 3주동안 시간이 남으면 사용아이템으로 확장하는ㄱ게 어떤가요?

SeRDic0705 commented on Assets/Scripts/Objects/Items/Item_KeepsakeData/KeepsakeSO.cs:
```diff
@@ -0,0 +1,12 @@
+using UnityEngine;
+
+[CreateAssetMenu(menuName = "ItemData/KeepsakeData", fileName = "Keepsake_")]
+public class KeepsakeSO : ItemData
```
Keepsake는 자주 사용하지 않는 단어라서 어떤 역할을 하는지 검색해보고 알았어요

조금 더 보편적인 Memorial이나 Souvenir정도가 더 직관적인 네이밍 같아요

jhwoo944 replied:
Memorial로 바꾸겠습니다

SeRDic0705 commented on Assets/Scripts/Objects/Items/Item_TicketData/TicketSO.cs:
```diff
@@ -0,0 +1,14 @@
+[CreateAssetMenu(menuName = "ItemData/TicketData", fileName = "Ticket_")]
+public class TicketSO : ItemData
+{
+    public GachaType gachaType;
```
통상 픽업 구분하지 않고 사용할 수 있는 교환권이라 가챠타입을 구분할 필요는 없을 것 같아요

가챠타입을 구분할 필요가 없으니 ScriptableObject로 구현할 필요도 없고, 그냥 Token에 통합시키는게 어떨까요?

jhwoo944 replied:
가챠에서 사용되는 티켓에 통상과 픽업의 구분이 없다는 뜻인가요?
보통은 통상 티켓이랑 픽업 티켓이랑 구분하지 않나요?
통상 티켓이랑 픽업 티켓을 구분하는 이유는 픽업 뽑기의 가치가 통상 뽑기보다 높으니까,
다이아 상점에서 픽업 티켓 패키지를 따로 판매한다던가 할 수도 있고
주간 컨텐츠나 기타 보상으로 통상 티켓을 주는 경우도 있고
블루 아카이브가 티켓 구분이 없는 특이 케이스이고, 
통상 뽑기와 픽업 뽑기를 나눌 거면 티켓도 두 종류로 나누는 게 요즘 가챠 게임의 주류 메타라고 생각하는데,
어떻게 생각하시나요?

SeRDic0705 replied:
이 역시 위의 이유와 동일합니다

물론 티켓가치의 변동을 주자는 의견에는 동의하지만

일단은 통일해두고 나중에 분리하는게 좋을 것 같아요

jhwoo944 replied:
> 이 역시 위의 이유와 동일합니다
> 
> 물론 티켓가치의 변동을 주자는 의견에는 동의하지만
> 
> 일단은 통일해두고 나중에 분리하는게 좋을 것 같아요

무슨 말씀 하시려는지 이해 했습니다.
그럼 가챠 티켓 분류는 통일 할게요.
근데 이미 만들어둔 Consumable, Ticket 아이템 타입, 그리고 관련 폴더 / 스크립트 / SO를 삭제하고 재작업하는 것도 비효율적인 것 같아요.  
일단 Consumable 타입은 당장 실제 아이템 / 기능 구현 까지는 하지 않되 파일, 스크립트, SO는 삭제하지 말고 그냥 두고, 
Ticket도 Token이랑 합칠 게 아니라 가챠 티켓, 던전 입장권 등의 티켓류 아이템을 분류하는 타입으로 두는 게 어떨까요?
Token 타입은 상점/특수 상점에서 재화를 교환하는 골드와 유사한 개념의 재화 타입을 염두에 두고 설계한 건데, 
여기에 가챠에서만 사용되는 가챠 티켓을 한 그룹으로 묶는 건 가챠 티켓 혼자 성격이 달라서 적합하지 않은 것 같아요.

SeRDic0705 replied:
네 티켓은 따로 두는게 좋을것 같네요.그렇게 진행해주세요 :)

SeRDic0705 commented on Assets/Scripts/Objects/Items/Item_TokenData/TokenSO.cs:
```diff
@@ -0,0 +1,13 @@
+[CreateAssetMenu(menuName = "ItemData/TokenData", fileName = "Token_")]
+public class TokenSO : ItemData
+{
+    public int eventId; //�̺�Ʈ ID
```
이벤트상점에서만 사용하려는 것으로 보이는데, 일반상점(pvp보상으로 획득하는 코인이라든가?)같은 경우도 고려하면

범주를 늘려서 "교환용 재화"로 설정하고 위의 가챠티켓도 여기에 넣어버리는게 어떨까요?

jhwoo944 replied:
굳이 이벤트 상점 뿐만 아니라 상점 코인이든 뭐든 다 포괄하는 의미에서 Token = 교환용 재화로 생각하고 만들었습니다
이 점은 설명이 부족했던 것 같네요

SeRDic0705 commented on Assets/Scripts/Objects/Items/Item_characterData/RankUp_characterData/RankUp_characterSO.cs:
```diff
@@ -0,0 +1,17 @@
+using static Constants;
+
+[CreateAssetMenu(menuName = "ItemData/character/RankUp_characterData", fileName = "RankUp_character_")]
+public class RankUp_characterSO : ItemData
```
더미데이터라도 조각과 완전히 기능이 같다면 그냥 지우는게 혹시 모를 오류 방지를 위해 좋을 것 같아요

SeRDic0705 commented on Assets/Scripts/Objects/Items/Item_equipData/ExpUp_equipData/ExpUp_equipSO.cs:
현재 기획안 상으로는 장비에 따로 레벨이 있거나, 방어구 무기별 강화재료가 따로 있는 것이 아니라

- 캐릭터의 레벨이 장비 강화 요구레벨에 도달할 때마다
- 해당 장비에 정해진 재료를 소모하여 ( 금속갑옷이라면 철 + 천, 신발이라면 가죽+천 등) 강화하고
- +4강에서 강화하면 다음 티어 장비로 넘어가는 방식입니다.

기획서에 작성을 했는데, 다시 보니 제 설명이 부족했던 것 같네요 죄송합니다.

장비 강화관련 아이템은 Material로 통일하고, 장비SO에 강화에 필요한 아이템과 수량을 Dictionary로 넣을 수 있을까요?

jhwoo944 replied:
장비 SO에 장비 티어 / 장비 강화 필요 재화 및 수량을 추가하고
아이템 SO에 장비 관련 아이템 타입을 장비 강화 재료로 통일해서 재작업하겠습니다

jhwoo944 replied:
티어 분류는 1티어 2티어 3티어 이렇게 숫자식인가요?
아니면 ItemRank를 따라서 나누나요?

*일단 ItemRank로 작업할게요

SeRDic0705 replied:
아이템 티어를 구분할 생각은 없긴했는데 들어보니 레어도정도는 표시해주는게 좋을것 같네요

일단 생각했던건 아이템 요구레벨 별로 강화재료가 다르고, 각 아이템이 다음 아이템의 id를 연결리스트처럼 들고 있는 방식입니다.


아이템 강화에는 다음과 같은 아이템들이 필요합니다.


튼튼한 검 +4   강철20개 고급촉매15개
강철검   인어의눈물4개
강철검+1   인어의눈물8개
강철검+2   인어의눈물12개
강철검+3   인어의눈물16개
강철검+4   인어의눈물20개
지배자의검   미스릴25개 최고급촉매20개
지배자의검+1   용의눈물5개
지배자의검+2   용의눈물10개


예시로 들기위해 대충 이름이랑 수량 생각해낸거니까 어떤 흐름인지 참고정도로 봐주세요.

강철검은 강철검+1의 id를 가지고 있고, 강화를 진행하면 강철검+1의 id를 호출해 장착하는거죠


생각해보니 이런 방식이면 고유무기 구현에 지장이 생길것같은데, 고유무기역시 필수요소가 아니기때문에 확장성에 대해서는 나중에 생각하도록하죠

jhwoo944 replied:
일단 현재 구현 중인 내용 자체는 설명하신 것과 거의 유사한 구조입니다.
CharacterSO에 장비 배열을 추가했고, 이 장비 배열에 티어별 장비 SO를 인덱스 순서대로 설정해두면 됩니다.
CharacterGrowth에 int 값 장비 티어를 추가했고, 
weapon = character.characterData.weapon[weaponTier]
이 코드대로 characterGrowth의 weapon값은 장비 배열에서 티어에 맞는 장비를 참조합니다.
강화 메서드는 장비의 강화 수치가 +4 이상이 될 경우 장비 티어 값을 +1 하고 강화 수치를 0강으로 초기화합니다.
이런 식으로 티어별 장비 구간은 다음과 같아요.

캐릭터 레벨 1 : 커먼 등급 장비을 장착한 채로 시작.

1~30레벨
커먼

캐릭터 레벨 30 : 커먼+4강에서 1단계 더 강화 시 → 레어 등급 장비

30~50레벨
레어

캐릭터 레벨 50 : 레어+4강에서 1단계 더 강화 시 → 에픽 등급 장비

50~70레벨
에픽

캐릭터 70레벨 : 에픽+4강에서 1단계 더 강화 시 → 레전드 등급 장비

70~90레벨
레전드

캐릭터 90레벨 : 레전드+4강에서 특별한 조건 만족 시 → 신화(전용장비. 구현 예정 없음.)

이대로 계속 작업할게요.

jhwoo944 committed:
SRPG-29 Fix: 버그 수정

 - CharacterGrowth.cs
1.
 - maxLevel get 부분 : if (Limit > 1) →  if (Limit >= 1)
2.
 - Exp set
 - level → Level

jhwoo944 committed:
SRPG-56 Fix: 주석 추가

 - AccountData에 주석추가

jhwoo944 committed:
Merge branch 'SRPG-29-AccountData' into SRPG-56-ItemBase

jhwoo944 committed:
SRPG-56 ItemBase

 - 장비 강화 구조 개편.
 - 캐릭터 SO는 장비 SO타입의 '장비 타입 / 성장 장비 배열을 가짐.
 - 예시) 인덱스 0 = common등급 한손검, 1 = rare등급 한손검, 2 = epic 한손검...
 - 장비 SO는 강화에 필요한 아이템 및 갯수 딕셔너리를 가짐.
 - 캐릭터 Growth는 무기 강화도 / 장비 강화도 필드를 가짐.
//
 - 캐릭터 랭크업 아이템 타입 / 장비 경험치 아이템 타입 삭제.
 - keepsake => mamorial로 rename.
//
 - 여러가지 객체를 참조하는 값을 int ID값으로 바꾸는 요청에 대해:
인지하고 있으나 당장 DB가 없어서 id값만으로는 해당 객체를 탐색 및 접근하지 못함 => 해당 객체의 세부 필드에 접근해야하는 코드를 작성하는데 불편함.
//
객체 스프라이트를 Load 형식으로 하는 건에 대해:
SO에 스프라이트 경로를 어떤 형식의 필드로 저장하고 Load하는지 구조가 감이 안잡혀서, 조언 및 의견 주시면 감사하겠습니다.
string으로 경로값을 저장하도록 해두면 될까요?

//
Todo: 아이템을 사용하는 기능들이 실제로 아이템을 갖고 있는지 인벤토리에서 확인하고 있다면 차감하도록 메서드 작성하기

jhwoo944 commented:
피드백 받은 대로 재작업했습니다.

1. 장비 강화 구조 개편.
 - 캐릭터 SO :  장비 SO타입 배열 추가 (이 캐릭터가  각 티어별로 어떤 장비를 끼는지의 정보를 저장함)
 - 예시) 인덱스 0 = common등급 한손검, 1 = rare등급 한손검, 2 = epic 한손검...
 - 장비 SO : 강화에 필요한 아이템 및 갯수(int 장비 ID, int 아이템 갯수) 딕셔너리 추가.
 - 캐릭터 Growth: 무기 레벨 및 방어구 레벨 → 무기 강화 수치 및 장비 강화 수치로 변경.

//

2. SO 삭제 및 이름 변경
 - 캐릭터 랭크업 아이템 타입 / 장비 경험치 아이템 타입 삭제.
 - 타입명 keepsake → mamorial로 일괄 변경

//

3. 게임 객체(itemData 등)을 참조하는 값을 int ID값로 변경해 달라는 요청에 대해:
인지하고 있으나 당장 DB 자체가 없어서 id값만으로는 객체의 세부 필드/메서드에 탐색 및 접근하지 못하기 때문에
세부 필드에 접근해야하는 코드를 작성하는데 불편이 있어서 일단은 그냥 둔 채 작업하고 있습니다.
유니티에서 DB를 연결해서 사용해본 경험이 없기도 하고, 아직 어떤 종류의 DB를 사용하는지도 안 정해진 것 같아서 
어떤 식으로 데이터에 접근하도록 코드를 작성해야할지 감이 안 잡히는데 이 부분에 대해 추가 의견 / 조언 /피드백 남겨주시면 감사하겠습니다.

//

4. 객체 스프라이트를 Load 형식으로 변경해달라는 요청에 대해:
이것도 역시 인지하고 있으나 SO에 스프라이트 경로를 어떤 형식의 필드로 저장할지 잘 모르겠습니다.
string으로 변수를 하나 선언해서 스프라이트가 위치한 Resource 경로값을 저장하도록 해두면 될까요?

//

5. 가챠 티켓 분류 및 Token의 범위에 대해:
Token 아이템 타입은 이벤트 재화 뿐만 아니라 PvP 코인 등 교환용 재화 전체를 포괄하는 의미로 의도하고 설정한 것이 맞습니다.
가챠 티켓의 타입은 픽업 티켓과 통상 티켓으로 종류를 나누어 놓는 게 여러모로 좋을 것 같고 또 실제로도 대다수의 가챠 게임에서 가챠에 사용할 수 있는 티켓의 종류를 구분하여 사용하고 있는데, 이 부분에 대해 어떻게 생각하시는지 의견 남겨주세요.

//

6. 교환권 / 소비품 개념의 아이템 타입에 대해:
시간에 따라 리젠되는 Ap 개념이 있는 게임인 이상 유저가 Ap를 비축해두고 자율적으로 관리 운용할 수 있도록 가능성을 열어 두는 편이 편의성을 보장하고 소소하게 게임의 재미를 더해주기 때문에, 사용하면 AP를 충전해주는 아이템은 반드시 있는게 맞다고 생각합니다.
사용 시에 아이템 / 재화를 지급하는 교환권류 아이템 타입군이 있는 게임도 있고, 바로 지급하거나 우편함으로 지급하는 게임도 있고, 3가지 형식을 모두 사용하는 게임도 있습니다.
그런데 Ap 충전 아이템을 만든다면 굳이 Ap 충전 아이템 타입으로 분리할 게 아니라 교환권류 아이템 타입으로 지정해 두는 편이 확장성 면에서 좋을 것 같아요.
인게임 전투에서 아이템(캐릭터 체력 회복제, 버프 물약 등)이 있는지 없는지 여부에 대해서도 의견 남겨 주세요.
저는 인게임 전투중에 사용할 수 있는 아이템은 없는게 좋다고 생각합니다.

//

7. Todo
아이템을 사용하는 메서드들(캐릭터 레벨업, 돌파, 장비 및 스킬 강화 등)에서, 
실제로 아이템을 갖고 있는지 인벤토리에서 확인하고,  있다면 차감하도록 메서드 작성하기
//

8. 스크립터블 오브젝트는 인스펙터에서 딕셔너리를 지원하지 않습니다.
장비 / 돌파 /  스킬 / 호감도 등 강화할 수 있는 요소들의 SO에 
딕셔너리<강화에 필요한 아이템 ID/필요한 아이템 갯수> 필드를 추가하고 그에 맞춰 강화 메서드를 작업했는데,
실제 SO에서 딕셔너리 키&값을 설정할 수 없는 문제가 생기네요.
어떻게 처리해야 좋을까요?

chai227chai committed:
feat & refactor : 디자인 변경과 로직 수정 중

GeCrabKani commented:
4. 객체 스프라이트를 Load 형식으로 변경

- 필요한 곳에서 직접 로드하는 방식
- 함수로 들고 있는 방식

이 있을 것 같습니다.

필요한 곳에서 직접 Managers.Resourca.Load<Sprite>($"{객체id}");로 호출을 하거나
함수로 public Sprite GetSprite() {Managers.Resourca.Load<Sprite>($"{객체id}");}를 작성하여 객체.GetSprite()로 부르는 방법이 있을 것 같습니다.
GetSprite() 함수를 만드는 것이 더 좋아보입니다.

5. 가챠 게임을 많이 해보진 않았으나 현재 "세븐나이츠 키우기"를 하고 있고 여기에서는 가챠 티켓 종류를 구분하고 있기는 합니다.

6. 어떨지 아직은 잘 모르겠습니다. 관련 코드를 적어나가면서 불편함이 있다면 바꿔도 좋을 것 같습니다.
인게임 내 물약의 경우 저는 없는 편이 좋다고 생각됩니다. 회복 포션이 있으면 플레이 타임이 길어져서 재미가 떨어질 것 같습니다.

7. SO의 문제가 아닙니다. 딕셔너리는 인스펙터에서 설정할 수 없습니다.
코드에서 설정을 해주어야 합니다.
코드 내에서 DB를 통해 값을 구해오고 이를 딕셔너리에 추가하면 됩니다.
현재 DB가 없는 상태에서 테스트를 하고 싶다면
- 직접 코드 내에서 데이터 생성 후 추가
- 리스트 선언 후 인스펙터에서 데이터 설정 & 코드에서 리스트 -> 딕셔너리로 데이터 이동
방법이 있습니다.

jhwoo944 committed:
SRPG-56 장비 구조 변경 및 장비강화 메서드 추가

 - 장비 및 장비 강화의 구조를 피드백 받은 내용대로 개편했습니다.
 - 각 SO의 스프라이트 필드를 제거했습니다.
 - 뽑기 티켓 분류를 통일했습니다.
 - AccountData의 ItemData 자료형을 <int, int>로 변경했습니다.
 - CharacterGrowthManager에 장비 강화 메서드를 추가했습니다.
 - CharacterGrowth에 장비 강화 관련 새 필드를 추가했습니다.

Todo: 스킬 레벨업 / 캐릭터 레벨업 메서드 구현

chai227chai committed:
refactor : 코드 변경 중

chai227chai committed:
refactor : 전투 로직과 애니메이션 코드 대폭 수정

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-56-ItemBase

jhwoo944 committed:
-

호감도 랭크업 메서드 주석처리

jhwoo944 committed:
Merge remote-tracking branch 'origin/SRPG-25_BattleSystem/RefactoringTest' into SRPG-56-ItemBase

jhwoo944 committed:
SRPG-56 성장 메서드가 아이템을 차감하도록 변경

 - CharacterGrowth 브랜치와 영역이 겹치는 작업을 ItemBase 브랜치에서 통합해서 작업했습니다.
필요한 아이템의 정보를 받아오고, 인벤토리에 해당 아이템이 있는지 확인하고, 사용 시 갯수를 차감하는 로직을 만들고 CharacterGrowth 메서드의 캐릭터 돌파, 장비 강화 메서드에 적용했습니다.

 - characterGrowthManager를 이제 Managers가 아니라 Character가 보유하고 있도록 변경했습니다.
Todo : 일단 기능이 정상적으로 돌아가게 하기 위해서 character와 characterGrowthManager가 상호 참조하고, 또
character와 characterGrowth도 상호 참조하도록 설정했는데, 상호 참조를 안 하면서 세 클래스를 연결하려면 어떻게 해야할 지 방법을 찾아봐야 할 것 같습니다.
//
 - 장비 레벨업 메서드
기획안에 나온 대로 조건을 체크하고 조건에 따라 장비 강화도를 상승 시키거나 티어를 한 단계 올립니다.

 - 캐릭터 돌파 메서드
캐릭터의 조각 아이템의 itemId 값이 캐릭터SO의 characterId값과 동일하다고 가정하고, 조건에 따라 캐릭터 돌파(Star 등급 Up) 또는 한계 돌파를 진행합니다.

//
Todo. CharacterGrowth
1. 캐릭터에게 Exp 상승 아이템을 사용하여 레벨업 시키는 기능 및 로직을 구현해야 합니다.
일단 캐릭터 Exp 값 자체는 상승하면 레벨이 올라가도록 설정해 두었지만 일반적으로 캐릭터 육성 창에서 진행되는 캐릭터 레벨업 구조
(예를 들어 블루아카이브: 캐릭터 육성 창- 등급별 캐릭터 경험치 아이템 4개 중에서 사용할 아이템을 고르고 갯수를 선택하기 - 확인 버튼으로 그 아이템들의 경험치 합계값을 캐릭터 Exp에 더하고 아이템들을 인벤토리에서 차감)
을 구현하려면 조금 복잡해질 것 같습니다.

2. 캐릭터 스킬 레벨업 기능을 구현해야 합니다.
일단 레벨업에 필요한 아이템과 아이템 갯수를 저장하는 딕셔너리가 있어야 합니다.
장비는 티어가 상승하면 상위 티어의 장비 SO를 적용 시키고 그 각각의 장비 SO에 강화에 필요한 아이템 딕셔너리가 저장되어 있는 방식이라 간단했는데, 스킬SO는 쭉 같은 스킬 SO를 사용 하지만 스킬 레벨에 따라서 요구 아이템 딕셔너리를 따로 적용 시켜줘야 하고, 그걸 또 진영 별로 나눠야 하니 딕셔너리 갯수가 진영 갯수 * 스킬 레벨 갯수만큼 필요해집니다. 이 딕셔너리 데이터를 어디에 저장시켜두면 좋을 지 고민입니다.
//
비쥬얼 스튜디오 인코딩 설정을 UTF-8로 바꿨는데 잘 적용되는지는 확인이 필요합니다.

jhwoo944 committed:
Merge remote-tracking branch 'origin/develop' into SRPG-56-ItemBase

jhwoo944 committed:
SRPG-40 CharacterGrowthManager 캐릭터 육성 관련 메서드 추가

 - 캐릭터 스킬 레벨업 메서드에서,
해당 스킬의 레벨업에 필요한 아이템 목록을 확인하고 차감하도록 변경했습니다.
 - 스킬 레벨업에 필요한 아이템 딕셔너리 데이터는 따로 저장할 곳이 마땅치 않고 DB도 아직 없는 관계로 따로 스태틱 클래스 및 딕셔너리를 만들어서 그 곳의 데이터를 가져와 체크하는 식으로 구성했습니다.
추후에 스킬 레벨업에 필요한 아이템 목록 DB를 구축한다면 그 데이터를 불러오도록 수정할 필요가 있습니다.

 - 스킬 레벨업에 필요한 진영별/레벨별 아이템 목록 데이터 딕셔너리를 저장하는 RankUp_skillMaterials 클래스 추가.

Todo : 캐릭터에 경험치 아이템을 사용해서 레벨업 시키는 로직 설계 및 기능 구현
